### PR TITLE
FAPI2 profile support

### PR DESCRIPTION
**Pull request by paulswartz**
Ref: https://github.com/erlef/oidcc/discussions/316

## Open Questions
- [x] should this split the profile atoms into `fapi2_security_profile` and `fapi2_message_signing`? The latter seems to also require request objects and JARM (even if the conformance suite doesn't seem to).

## TODO
- [x] test coverage
- [x] documentation
- [x] Elixir implementation

### CODE REVIEW

**Comment by maennchen** [comment]
> should this split the profile atoms into fapi2_security_profile and fapi2_message_signing? 

That would probably be good. FAPI1 also has very different requirements between the specs.

File: src/oidcc_profile.erl [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+-include("oidcc_provider_configuration.hrl").
+
+-export([apply_profiles/2]).
+
+%% @private
+-spec apply_profiles(ClientContext, Opts) ->
+    {ok, ClientContext, Opts} | {error, oidcc_client_context:error()}
+when
+    ClientContext :: oidcc_client_context:t(),
+    Opts :: map().
```
Comment (maennchen) [inline src/oidcc_profile.erl:20]: I believe we should define a type for the opts map and a specific type for the profile atoms.

Comment (paulswartz) [reply src/oidcc_profile.erl:20]: Should that live in `oidcc_client_context` or `oidcc_profile`?

Comment (maennchen) [reply src/oidcc_profile.erl:20]: We‘ve located the types close to the code so far. (Which resulted in public modules where only the types are public but all fns are private.)

I think we should stick with that for now. (=> `oidcc_profile`)

**Comment by maennchen** [comment]
@paulswartz mTLS should be relatively simple to handle:

The user can already supply `ssl` options. If you include `cert[file]`, `key[file]` and `verify: :verify_peer`, mTLS is enabled.

So we can just verify that they either are set and otherwise require DPoP.

File: lib/oidcc/client_context.ex [hunk 2]
```diff
@@ -139,6 +139,44 @@ defmodule Oidcc.ClientContext do
...
+      ...>   )
+      ...>
+      ...> {:ok, %Oidcc.ClientContext{}, %{}} =
+      ...>   Oidcc.ClientContext.apply_profiles(
+      ...>     client_context,
+      ...>     %{profiles: [:fapi2_message_signing]}
+      ...>   )
+  """
+  @doc since: "3.2.0"
+  @spec apply_profiles(t(), map()) :: {:ok, t(), map()} | {:error, :oidcc_client_context.error()}
```
Comment (maennchen) [inline lib/oidcc/client_context.ex:169]:
```suggestion
  @spec apply_profiles(t(), :oidcc_profile.opts()) :: {:ok, t(), :oidcc_profile.opts_no_profiles()} | {:error, :oidcc_client_context.error()}
```

**Review by maennchen — CHANGES_REQUESTED** [review]

**Comment by maennchen** [comment]
After this minor correction good for merge.

**Comment by maennchen** [comment]
@paulswartz Can you open issues for all the TODOs that are missing for full compliance? (Like JARM and mTLS)

**Review by maennchen — APPROVED** [review]

**Comment by maennchen** [comment]
@paulswartz Can you run `mix format` so that I can merge?

**Review by maennchen — APPROVED** [review]

**Pull request merged**